src/util/yp_newline_list.c

```diff
@@ -0,0 +1,117 @@
+#include "yarp/util/yp_newline_list.h"
+
+// Initialize a new newline list with the given capacity. Returns true if the
+// allocation of the offsets succeeds, otherwise returns false.
+bool
+yp_newline_list_init(yp_newline_list_t *list, const char *start, size_t capacity) {
+    list->offsets = (size_t *) calloc(capacity, sizeof(size_t));
+    if (list->offsets == NULL) return false;
+
+    list->start = start;
+
+    // This is 1 instead of 0 because we want to include the first line of the
+    // file as having offset 0, which is set because of calloc.
+    list->size = 1;
+    list->capacity = capacity;
+
+    list->last_index = 0;
+    list->last_offset = 0;
```

Comment:
Not sure where I land on this, but as is, we're missing the optimization on the very first line. We could consider setting it to `-1` or something so we keep the optimization there. Again though, not sure if the signed int is worth getting this extra optimization...

Reply:
Yeah I figured it was fine to lose it at the beginning honestly.